File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
    def subscribed
      stream_from "chat_channel"
    end
  
    def unsubscribed
      stop_all_streams
    end
    
    def send_message(data)
      user = User.find_by_id(data["user_id"])
      if user.present?
        message = Message.create(content: data["content"], user: user)
        rendered_message = render_message(message)
        puts "Rendered message: #{rendered_message}"
        ActionCable.server.broadcast("chat_channel", message: rendered_message, user_id: user.id, receiver_id: data["receiver_id"])
      else
        puts "User not found with ID: #{data["user_id"]}"
      end
    end
  
    private
  
    def render_message(message)
     message.content
    end
  end
  

File: client/src/components/javascript/channels/chat_channel.js
import { createConsumer } from '@rails/actioncable';

const getWebSocketURL = () => {
  return 'ws://localhost:3000/cable';
};

const consumer = createConsumer(getWebSocketURL());

const sendMessage = (channel, content, user_id, receiver_id) => {
  channel.perform('send_message', { content, user_id, receiver_id });
};


const createChatChannel = (receivedMessageCallback, user_id, receiver_id) => {
  console.log('Creating chat channel');
  
  const channel = consumer.subscriptions.create(
    {
      channel: 'ChatChannel',
      user_id: user_id,
      receiver_id: receiver_id
    },
    {
      connected() {
        console.log('Connected to chat channel');
      },

      disconnected() {
        console.log('Disconnected from chat channel');
      },

      received(data) {
        console.log("Received data:", data);
        if (data.message && data.message.trim() !== "") {
          receivedMessageCallback({
            content: data.message.trim(),
            user_id: data.user_id,
            receiver_id: data.receiver_id,
          });
        } else {
          console.warn("Received empty message or message is not defined");
        }
      },
      

      send_message(content, user_id, receiver_id) {
        this.perform('send_message', { content, user_id, receiver_id });
      },
    }
  );

  return {
    sendMessage: (content, user_id, receiver_id) => channel.send_message(content, user_id, receiver_id),
    unsubscribe: () => consumer.subscriptions.remove(channel),

  };
};


export default createChatChannel;
